lib/rubocop/cop/lint/redundant_with_object.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for redundant `with_object`.
+      #
+      # @example
+      #   # bad
+      #   ary.each_with_object([]) do |v|
+      #     v
+      #   end
+      #
+      #   # good
+      #   ary.each do |v|
+      #     v
+      #   end
+      #
+      #   # bad
+      #   ary.each.with_object([]) do |v|
+      #     v
+      #   end
+      #
+      #   # good
+      #   ary.each do |v|
+      #     v
+      #   end
+      #
+      class RedundantWithObject < Cop
+        MSG_EACH_WITH_OBJECT = 'Use `each` instead of `each_with_object`.'
+                               .freeze
+        MSG_WITH_OBJECT = 'Remove redundant `with_object`.'.freeze
+
+        def_node_matcher :redundant_with_object?, <<-PATTERN
+          (block
+            $(send _ {:each_with_object :with_object}
+              (_))
+            (args
+              (arg _))
+            ...)
+        PATTERN
+
+        def on_block(node)
+          redundant_with_object?(node) do |send|
+            add_offense(node, with_object_range(send))
+          end
+        end
+
+        def autocorrect(node)
+          lambda do |corrector|
+            redundant_with_object?(node) do |send|
+              if send.method_name == :each_with_object
+                corrector.replace(with_object_range(send), 'each')
+              else
+                corrector.remove(with_object_range(send))
+                corrector.remove(send.loc.dot)
+              end
+            end
+          end
+        end
+
+        private
+
+        def message(node)
+          if node.method_name == :each_with_object
+            MSG_EACH_WITH_OBJECT
+          else
+            MSG_WITH_OBJECT
+          end
+        end
+
+        def with_object_range(send)
+          range_between(
+            send.loc.selector.begin_pos,
+            send.source.length
```

Comment:
The end range is different than the `with_index` cop. I saw that this was changed from `send.loc.selector.end_pos + 1` when tests were added for the method call without parentheses. Does this need to be changed in the `with_index` cop?

Reply:
Current `Lint/RedundantWithObject` cop does not check the argument with_index. This is an oversight of me, in fact it would be better to handle the case with the `with_index` argument. It can be improved as an implementation of `Lint/RedundantWithObject` cop.